src/main/java/org/openrewrite/java/testing/junit5/UpdateMockWebServer.java

```diff
@@ -0,0 +1,159 @@
+
+import java.util.Collections;
+import java.util.UUID;
+
+/**
+ * Recipe for converting JUnit4 okhttp3 MockWebServer Rules with their JUnit5 equivalent.
+ * Note this recipe upgrades okhttp3 to version 4.x there are a few backwards incompatible changes: https://square.github.io/okhttp/upgrading_to_okhttp_4/#backwards-incompatible-changes
+ * - If MockWebServer Rule exists remove the Rule annotation and update okhttp3 to version 4.x
+ * - If AfterEach method exists insert a close statement for the MockWebServer and throws for IOException
+ * - If AfterEach does not exist then insert new afterEachTest method closing MockWebServer
+ */
+public class UpdateMockWebServer extends Recipe {
+    private static final AnnotationMatcher RULE_MATCHER = new AnnotationMatcher("@org.junit.Rule");
+    private static final AnnotationMatcher AFTER_EACH_MATCHER = new AnnotationMatcher("@org.junit.jupiter.api.AfterEach");
+
+    private static final ThreadLocal<JavaParser> OKHTTP3_PARSER = ThreadLocal.withInitial(() ->
+            JavaParser.fromJavaVersion().dependsOn(Collections.singletonList(
+                    Parser.Input.fromString("package okhttp3.mockwebserver;" +
+                            "public final class MockWebServer extends java.io.Closeable {}"
+                    )
+            )).build());
+
+    @Override
+    public String getDisplayName() {
+        return "okhttp3 3.x MockWebserver @Rule To 4.x MockWebServer";
+    }
+
+    @Override
+    public String getDescription() {
+        return "Replace usages of okhttp3 3.x @Rule MockWebServer with 4.x MockWebServer.";
+    }
+
+    @Override
+    protected @Nullable TreeVisitor<?, ExecutionContext> getApplicableTest() {
+        return new JavaIsoVisitor<ExecutionContext>() {
+            @Override
+            public J.CompilationUnit visitCompilationUnit(J.CompilationUnit cu, ExecutionContext executionContext) {
+                J.CompilationUnit c = super.visitCompilationUnit(cu, executionContext);
+                if (!FindTypes.find(cu, "org.junit.Rule").isEmpty()
+                        && !FindTypes.find(cu, "okhttp3.mockwebserver.MockWebServer").isEmpty()) {
+                    c = c.withMarker(new RecipeSearchResult(UpdateMockWebServer.this));
+                }
+                return c;
+            }
+        };
+    }
+
+    @Override
+    protected TreeVisitor<?, ExecutionContext> getVisitor() {
+        return new JavaIsoVisitor<ExecutionContext>() {
+            private final static String MOCK_WEBSERVER_RULE = "mock-web-server-rule";
+            private final static String AFTER_EACH_METHOD = "after-each-method";
+
+            @Override
+            public J.ClassDeclaration visitClassDeclaration(J.ClassDeclaration classDecl, ExecutionContext executionContext) {
+                J.ClassDeclaration cd = super.visitClassDeclaration(classDecl, executionContext);
+                String mockWebServerVarableName = getCursor().pollMessage(MOCK_WEBSERVER_RULE);
```

Comment:
Mostly, calling varable names out for reminiscing.

Reply:
nice catch, thanks @traceyyoshima